grails-app/controllers/org/pih/warehouse/core/DocumentController.groovy

```diff
@@ -207,10 +210,17 @@ class DocumentController {
                     contentType: file.contentType,
                     extension: FileUtil.getExtension(file.originalFilename),
                     documentNumber: command.documentNumber,
-                    documentType: DocumentType.get(typeId))
+                    documentType: documentType)
+
+            documentInstance.validate()
+
+            List<DocumentType> nonTemplateDocumentTypes = documentService.getNonTemplateDocumentTypes()
+            if (!nonTemplateDocumentTypes.contains(documentType)) {
+                documentInstance.errors.reject("documentType", "Template types are not allowed for this document upload")
+            }
```

Comment:
You're right I'm not a fan of having to do this :laughing: I'm assuming you're doing it to avoid needing more code change below where it's checking the domain validator errors. Pretty hacky hehe.

I'd think a more proper solution would be to add a custom `DocumentValidator` that wraps the existing validator on the domain and then also performs more complex checks like this one. Then we could just call `documentValidator.validate()` and it'll all be handled by a single call. That's not the simplest of changes though so if you don't want to do that right now can you add a todo here asking for a future refactor?

One thing to check. In the ticket it says "users should not be able to create new docs of this type. That should be limited to admins (prob superuser)." How are admins adding template documents currently? Is it via this API? If so, this change will prevent them from doing so

Reply:
I believe none of the users should be able to add template type of documents in these places (invoice, stockMovement, PO, etc... ), we have a dedicated page for adding all sorts of documents that each user can access so we are not limiting anything.
I believe the goal for this ticket was to prevent users from adding template documents on individual invoices, stock movements etc.. because they are accessible globally on other invoices and stockmovement and not limited to the one where it was uploaded.

Reply:
yes agreed, I just want to confirm that the general page for adding documents doesn't also use this flow. Can you verify that with your change you can still navigate to the /openboxes/document/list page and upload a template document?

Reply:
The add document pages/subpages that I implemented a change for use the 
`DocumentController` > `saveDocument` or `updateDocument`
while the document/list uses the
`DocumentController` > `save` or `update`
which I have not touched

Reply:
that's not confusing at all :laughing: thanks for confirming